### PR TITLE
Add API route to create many users

### PR DESCRIPTION
**Pull request by m5r**
# Description

This PR adds a `createManyUsers` function to the `usersService` that works in a similar way as the `createUser` function.  
It can be used directly from the service `usersService.createManyUsers([])` within the API or from a standalone API route `POST /api/v2/users`

This is part of the work in progress for CSV bulk users import

[Related cht-docs PR](https://github.com/medic/cht-docs/pull/614)

medic/cht-core#7490

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: api/tests/mocha/services/users.spec.js [hunk 1]
```diff
@@ -674,6 +674,176 @@ describe('Users service', () => {
...
+        );
+        chai.expect(error.failingIndexes[1].passwordError.message.translationKey).to.equal('password.weak');
+        chai.expect(error.failingIndexes[1].index).to.equal(1);
+
+        chai.expect(error.code).to.equal(400);
+      }
+    });
+
+    describe('errors at insertion', () => {
+      it('returns responses with errors if contact.parent lookup fails', async () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:758]: Please add tests where: 
- validateNewUsername fails
- createContact fails
- storeUpdatedPlace fails
- manageTokenLogin fails

**Review by dianabarsan — COMMENTED** [review]
Nice work! I really like how this is taking shape!
I left a couple of requests and suggestions inline.

File: api/src/routing.js [hunk 2]
```diff
@@ -390,6 +390,7 @@ app.post('/api/v1/forms/validate', textParser, forms.validate);
 
 app.get('/api/v1/users', users.get);
 app.postJson('/api/v1/users', users.create);
+app.postJson('/api/v2/users', users.createMany);
```
[the remark this answers is not in the document]
Comment (m5r) [reply api/src/routing.js:393]: I don't have any strong preference on whether it should replace the previous endpoint.  
I initially wanted to add a new v1 endpoint but naming that endpoint anything other than `/users` won't follow the REST naming convention.
I saw a handful of endpoints having a v2 (namely `/monitoring`, `/records` and thought I would use the same approach.

In any case, I will push changes to merge the two endpoints into `/api/v1/users` and keep it backwards compatible with the existing behavior

[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/routing.js:393]: I think `records` and `monitoring` have all v2 methods covered. 
Whereas `users` has : 

1. app.get('/api/v1/users', users.get);
2. app.postJson('/api/v1/users', users.create);
3. app.postJson('/api/v1/users/:username', users.update);
4. app.delete('/api/v1/users/:username', users.delete);

All of these endpoints would need their v2 variants. You can't just have v2 for `POST`, but not for GET, another POST and DELETE.

[the remark this answers is not in the document]
Comment (m5r) [reply api/src/routing.js:393]: Ah thanks, it makes sense now! I thought it meant having all endpoints of the whole API being upgraded to v2

File: api/src/services/users.js [hunk 3]
```diff
@@ -577,6 +577,117 @@ module.exports = {
...
+      for (const { tokenLoginError, index } of failingIndexes) {
+        errorMessge += `\nError ${tokenLoginError.msg} for user at index ${index}`;
+      }
+      const error = new Error(errorMessge);
+      error.code = 400;
+      error.failingIndexes = failingIndexes;
+      return Promise.reject(error);
+    }
+
+    const passwordErrors = users.map(user => validatePassword(user.password));
```
[the remark this answers is not in the document]
Comment (m5r) [reply api/src/services/users.js]: With usual username/password logins, an error is thrown before this line of code is executed If the password is not set  
And in the case of token logins, a password is randomly generated beforehand with `tokenLogin.validateTokenLogin()` so there will always be a password set before running this validation

File: api/src/services/users.js [hunk 4]
```diff
@@ -547,34 +601,127 @@ module.exports = {
...
+   * @param {string} users[].username
+   * @param {string=} users[].phone Not required if the password is provided.
+   * @param {string=} users[].password Not required if the phone number is provided.
+   * @param {string[]} users[].roles
+   * @param {(Object|string)=} users[].place Can either be a place object or an existing place id.
+   * @param {(Object|string)=} users[].contact Can either be a contact object or an existing place id.
+   * @param {string=} users[].type Deprecated. Used to infer user's roles
+   * @param {string} appUrl   request protocol://hostname
+   */
+  async createUsers(users, appUrl) {
```
Comment (dianabarsan) [inline api/src/services/users.js:649]: This is a very long function. Please break it down into smaller chucks.

File: api/tests/mocha/services/users.spec.js [hunk 5]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+          'The password is too easy to guess. Include a range of types of characters to increase the score.',
+        );
+        chai.expect(error.failingIndexes[1].passwordError.message.translationKey).to.equal('password.weak');
+        chai.expect(error.failingIndexes[1].index).to.equal(1);
+
+        chai.expect(error.code).to.equal(400);
+      }
+    });
+
+    it('returns error if one of the users has a missing phone number', async () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:779]:
```suggestion
it('returns error if one of the users has a missing phone number and should login by SMS', async () => {
```

Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:779]:
```suggestion
    it('returns error if one of the users has a missing phone number and should login by SMS', async () => {
```

File: api/tests/mocha/services/users.spec.js [hunk 6]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+        ]);
+        chai.assert.fail('Should have thrown');
+      } catch (error) {
+        chai.expect(error.code).to.equal(400);
+        chai.expect(error.failingIndexes[0].fields).to.deep.equal(['phone']);
+        chai.expect(error.failingIndexes[0].index).to.equal(1);
+      }
+    });
+
+    it('returns error if one of the users has an invalid phone number', async () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:808]:
```suggestion
    it('returns error if one of the users has an invalid phone number and should login by SMS', async () => {
```

File: api/tests/mocha/services/users.spec.js [hunk 7]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+          {
+            username: 'sally',
+            roles: ['a', 'b'],
+            phone: '+40 755 69-69-69',
+            token_login: true,
+          },
+          {
+            username: 'sally',
+            roles: ['a', 'b'],
+            phone: '123',
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:826]: Please use a valid phone number here.

File: api/tests/mocha/services/users.spec.js [hunk 8]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+          {
+            username: 'sally',
+            roles: ['a', 'b'],
+            token_login: true,
+          },
+        ]);
+        chai.assert.fail('Should have thrown');
+      } catch (error) {
+        chai.expect(error.code).to.equal(400);
+        chai.expect(error.failingIndexes[0].fields).to.deep.equal(['phone']);
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:803]: The test name says "'returns error if one of the users has a missing phone number", and your payload has 2 users, one with an invalid phone number (index 0) and one with a missing phone number (index 1). 
Your assertions are only treating index 0. 
Please change index 0 to have a valid phone number and assert that index 1 is throwing the 400.

File: api/tests/mocha/services/users.spec.js [hunk 9]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+        chai.expect(error.code).to.equal(400);
+        chai.expect(error.failingIndexes[0].tokenLoginError.msg).to.equal(
+          'A valid phone number is required for SMS login.'
+        );
+        chai.expect(error.failingIndexes[0].tokenLoginError.key).to.equal('configuration.enable.token.login.phone');
+        chai.expect(error.failingIndexes[0].index).to.equal(1);
+      }
+    });
+
+    it('should normalize phone number and change password (if provided)', async () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:841]: Please mention that token login is enabled in the test name.

File: api/tests/mocha/services/users.spec.js [hunk 10]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+      service.__set__('createContact', sinon.stub().resolves());
+      service.__set__('storeUpdatedPlace', sinon.stub().resolves());
+      service.__set__('createUserSettings', sinon.stub().resolves());
+      sinon.stub(places, 'getPlace').resolves({ _id: 'foo' });
+      userData.place = 'foo';
+      const response = await service.createUsers([userData]);
+      chai.expect(response).to.deep.equal([{}]);
+    });
+
+    it('succeeds and response contains the user, contact and user settings fields', async () => {
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:1143]: Please update this test so two users are inserted, so we see how the response looks like when multiple users are successfully created.

File: api/tests/mocha/services/users.spec.js [hunk 11]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+    it('errors if username exists in _users db', async () => {
+      const userData = {
+        username: 'x',
+        password: COMPLEX_PASSWORD,
+        place: { name: 'x' },
+        contact: { 'parent': 'x' },
+        type: 'national-manager'
+      };
+      sinon.stub(db.users, 'get').resolves('bob lives here already.');
+      sinon.stub(db.medic, 'get').resolves();
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:1218]: For the test to be correct, this should reject with a 404.

File: api/tests/mocha/services/users.spec.js [hunk 12]
```diff
@@ -674,6 +675,574 @@ describe('Users service', () => {
...
+
+    it('errors if username exists in medic db', async () => {
+      const userData = {
+        username: 'x',
+        password: COMPLEX_PASSWORD,
+        place: { name: 'x' },
+        contact: { 'parent': 'x' },
+        type: 'national-manager'
+      };
+      sinon.stub(db.users, 'get').resolves();
```
Comment (dianabarsan) [inline api/tests/mocha/services/users.spec.js:1235]: For the test to be correct, this should reject with a 404.

File: tests/e2e/api/controllers/users.js [hunk 13]
```diff
@@ -844,6 +844,102 @@ describe('Users API', () => {
           .then(() => expectPasswordLoginToWork(user));
       });
 
+      it('should create and update many users correctly w/o token_login', async () => {
```
Comment (dianabarsan) [inline tests/e2e/api/controllers/users.js:847]: Nice! 

I think we should have a similar test that creates users with token_login enabled, and one where some users fail to be created, so we can assert on the response we get e2e. 
Thanks!

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! Thanks for all the updates. 
I left a couple of suggestions and requests inline.

File: api/src/services/users.js [hunk 14]
```diff
@@ -104,10 +111,22 @@ const validateContact = (id, placeID) => {
   return db.medic.get(id)
     .then(doc => {
       if (!people.isAPerson(doc)) {
-        return Promise.reject(error400('Wrong type, contact is not a person.','contact.type.wrong'));
+        const msg = 'Wrong type, contact is not a person.';
+        return Promise.reject(error400(msg, {
```
[the remark this answers is not in the document]
Comment (m5r) [reply api/src/services/users.js:115]: That was my first thought but I was feeling having two different ways of calling the same function would be considered a code smell.  
I agree this alternative has its flaws but it's consistent at least.

I'm reverting this change and making the function act differently based on the arguments' type

File: api/tests/mocha/services/users.spec.js (hunk 7 above)
Comment (m5r) [reply api/tests/mocha/services/users.spec.js:826]: This test here asserts the phone number is invalid for the user at index 1. Is there a correct phone number format that is considered invalid by `google-libphonenumber`?

Comment (dianabarsan) [reply api/tests/mocha/services/users.spec.js:826]: Oh, so sorry, I was mislead by the [0]. It's great!

**Comment by m5r** [comment]
@dianabarsan please re-review, thanks!

File: api/src/services/users.js [hunk 15]
```diff
@@ -59,9 +60,20 @@ const illegalDataModificationAttempts = data =>
...
  */
-const error400 = (msg, key, params) => ({
-  code: 400, message: { message: msg, translationKey: key, translationParams: params }
-});
+const error400 = (msg, key, params) => {
+  const error = new Error(msg);
+  error.code = 400;
+
+  if (typeof key === 'string') {
+    Object.assign(error, {
```
Comment (dianabarsan) [inline api/src/services/users.js:68]: Can we return here and avoid the extra `else`?

File: api/src/services/users.js [hunk 16]
```diff
@@ -565,16 +619,70 @@ module.exports = {
...
+    const hasFailures = promises.some(promise => promise.status === 'rejected');
+    if (hasFailures) {
+      promises.forEach((promise, index) => {
+        if (promise.status === 'rejected') {
+          response[index].error = promise.reason.message;
+        }
+      });
+    }
+
+    return response;
```
Comment (dianabarsan) [inline api/src/services/users.js:685]: I we change `createUserEntities` to create and return the response object., this whole section can be rewritten like: 
```suggestion
    const promises = await allPromisesSettled(users.map((user) => createUserEntities(user, appUrl)));
    return promises.map(promise => {
      return promise.status === 'fulfilled' ? promise.value : { error: promise.reason.message };
    });
```

**Review by dianabarsan — COMMENTED** [review]
Very very nice work! 
I just left a couple of recommendations about readability.

File: api/src/services/users.js [hunk 17]
```diff
@@ -522,6 +534,48 @@ const validateUserContact = (data, user, userSettings) => {
...
+const validateUserFields = (users) => {
+  const missingFieldsFailingIndexes = [];
+  const tokenLoginFailingIndexes = [];
+  const passwordFailingIndexes = [];
+  users.forEach((user, index) => {
+    const fields = missingFields(user);
+    const hasMissingFields = fields.length > 0;
+    if (hasMissingFields) {
+      missingFieldsFailingIndexes.push({ fields, index });
+      return;
```
[the remark this answers is not in the document]
Comment (m5r) [reply api/src/services/users.js:559]: It makes sense to do that but in that case we must keep `createUser` behavior consistent and make it return all validation errors as well. Was there a specific reason why it was implemented that way then?

[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/users.js:559]: > Was there a specific reason why it was implemented that way then. 

No idea, was before I joined. Commit from 2016 first introducing this behavior: https://github.com/medic/cht-core/commit/e69c774b8ac3abd58bbfe3056208cac76fc0a6ad 

I added an issue to track this improvement: https://github.com/medic/cht-core/issues/7494

**Review by dianabarsan — APPROVED** [review]
Awesome!

**Pull request merged**